### PR TITLE
Block endpoints addition until all policies are processed from k8s

### PR DESCRIPTION
**Pull request by aanm**
For a large number of policies it is possible Cilium signalizes that caches
were synced with Kubernetes although they might not have been processed
and pushed down the local policy repository.

With the following diff is easy to replicate the bug:

```golang
diff --git a/daemon/k8s_watcher.go b/daemon/k8s_watcher.go
index da3fcba1e..3e6ddef6c 100644
--- a/daemon/k8s_watcher.go
+++ b/daemon/k8s_watcher.go
@@ -571,10 +571,13 @@ func (d *Daemon) EnableK8sWatcher(reSyncPeriod time.Duration) error {
                                metrics.EventTSK8s.SetToCurrentTime()
                                if cnp := copyObjToV2CNP(obj); cnp != nil {
                                        serCNPs.Enqueue(func() error {
+                                               time.Sleep(5 * time.Second)
+                                               log.Debugf("adding cilium network policy after sleeping")
                                                d.addCiliumNetworkPolicyV2(cnpStore, cnp)
                                                return nil
                                        }, serializer.NoRetry)
                                }
+                               log.Debugf("adding cilium network policy in AddFunc")
                        },
                        UpdateFunc: func(oldObj, newObj interface{}) {
                                metrics.EventTSK8s.SetToCurrentTime()
```
would give the following logs:

```
msg="cache synced" kubernetesResource=NetworkPolicy subsys=daemon
msg="cache synced" kubernetesResource=Service subsys=daemon
msg="cache synced" kubernetesResource=Endpoint subsys=daemon
msg="adding cilium network policy in AddFunc" subsys=daemon
msg="cache synced" kubernetesResource=CiliumNetworkPolicy subsys=daemon
msg="adding cilium network policy after sleeping" subsys=daemon
```

With this commit we make sure all policies are received and processed by
Cilium before signalizing the cache was synced with Kubernetes.

As we can see in the following logs, the policies were processed and
added to the policy repository before signalizing that caches were
synced.

```
msg="adding cilium network policy in AddFunc" subsys=daemon
msg="cache synced" kubernetesResource=NetworkPolicy subsys=daemon
msg="cache synced" kubernetesResource=Service subsys=daemon
msg="cache synced" kubernetesResource=Endpoint subsys=daemon
msg="adding cilium network policy after sleeping" subsys=daemon
msg="cache synced" kubernetesResource=CiliumNetworkPolicy subsys=daemon
```

Fixes: d440059 ("daemon: block until initial policy list")
Signed-off-by: André Martins <andre@cilium.io>

Fixes: https://github.com/cilium/cilium/issues/5038

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/5486)

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

File: daemon/k8s_watcher.go [hunk 1]
```diff
@@ -317,19 +317,24 @@ func k8sErrorHandler(e error) {
...
 		scopedLog := log.WithField("kubernetesResource", resourceName)
 		scopedLog.Debug("waiting for cache to synchronize")
-		if ok := cache.WaitForCacheSync(completed, informer.HasSynced); !ok {
+		if ok := cache.WaitForCacheSync(wait.NeverStop, informer.HasSynced); !ok {
 			// Fatally exit it resource fails to sync
 			scopedLog.Fatalf("failed to wait for cache to sync")
 		}
-		scopedLog.Debug("cache synced")
-		waitGroup.Done()
+		queue.Enqueue(func() error {
```
Comment (ianvernon) [inline daemon/k8s_watcher.go]: Add a comment indicating why we need to add this to the function queue. The reasoning is subtle, in that while the informer indicates that it has "synced", we want to make sure that Cilium has properly consumed said event, and only then can we indicate that the information has been processed by Cilium.

**Review by ianvernon — COMMENTED** [review]
Awesome find - I was not aware of this subtlety. Please add requested documentation.

File: daemon/k8s_watcher.go (hunk 1 above)
Comment (aanm) [reply daemon/k8s_watcher.go]: Done

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
all previous tests passed. just added a comment requested by @ianvernon triggering the ginkgo tests just to make sure nothing is broken

File: daemon/k8s_watcher.go [hunk 2]
```diff
@@ -317,19 +317,27 @@ func k8sErrorHandler(e error) {
...
+		if ok := cache.WaitForCacheSync(wait.NeverStop, informer.HasSynced); !ok {
 			// Fatally exit it resource fails to sync
 			scopedLog.Fatalf("failed to wait for cache to sync")
 		}
-		scopedLog.Debug("cache synced")
-		waitGroup.Done()
+		// Enqueue the waitGroup.Done to signalize the that all other functions
+		// queued were serialized. This will make sure the all events received
+		// after the cached was synced were processed by the queue.
+		queue.Enqueue(func() error {
```
Comment (ianvernon) [inline daemon/k8s_watcher.go:336]: Perfect , thanks for adding this.

**Review by ianvernon — APPROVED** [review]

**Review by tgraf — APPROVED** [review]

**Pull request merged**